Library/Homebrew/dev-cmd/audit.rb

```diff
@@ -897,7 +897,7 @@ def audit_text
     end
 
     def audit_lines
-      text.without_patch.split("\n").each_with_index do |line, lineno|
+      text.without_patch.gsub(/"\s*\\\n\s*"/, "").split("\n").each_with_index do |line, lineno|
```

Comment:
Can you explain the regex a bit more? I'm not quite seeing how it'll work.

Reply:
It deletes any text that starts with `"`, followed by any amount (including none) of whitespace, a literal `\` and a newline, and again any whitespace and a `"`.

So in the example code, as is in the formula (I’ve kept the indentation):

```ruby
      system "#{bin}/kafka-topics --zookeeper localhost:2181 --create --if-not-exists --replication-factor 1 " \
             "--partitions 1 --topic test > #{testpath}/kafka/demo.out 2>/dev/null"
```

The regex will match and delete the end of the first line and beginning of the second:

```
 " \
             "
```

Essentially transforming both lines into one with correct code.

Reply:
So because the text is all in one String at the moment, with the newlines embedded in it, it's possible to look for this sequence and remove it:

- double-quote
- some (optional spaces)
- a **backslash** (i.e. the continuation character)
- a newline
- some more (optional) spaces
- another double-quote.

It doesn't try and concatenate every continuation line, only the ones with double-quotes each side of the continuation character so it can remove the pair of double quotes. I thought it best to limit the scope.

My before and after test was `brew audit --strict kafka` .

I thought about running `brew audit --strict`on all the formulae but this takes over a day on my computer. So I ran it on a couple of others that used 'system' (zsh and tomcat) and they seemed OK. But I could run them all if you like. (I have an output from a week or two ago with the results of a full audit run to compare with.)

Instead of Regex, I considered changing `Formula#system` to allow the use of the parameterised version of Kernel#system with its file redirection arguments (see [Kernel#spawn](https://ruby-doc.org/core-2.6.4/Kernel.html#method-i-spawn)) but that seemed hard work for this sort of minor, style type change.

Reply:
Hmm, that doesn't seem very robust given the possible different ways of creating strings in Ruby.

What would be a better call (and you might be game for this given your username) is porting this (legacy) audit to be one of the newer RuboCop ones: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/rubocops/text.rb

They are a bit more of a pain to write but that will get you a Proper Ruby Parser which will make it more robust.

Reply:
Do [`chisel`](https://github.com/Homebrew/homebrew-core/blob/ef65542ed2b9899ec9f0a79759155705ae178e99/Formula/chisel.rb#L26-L27) and [`liblcf`](https://github.com/Homebrew/homebrew-core/blob/ef65542ed2b9899ec9f0a79759155705ae178e99/Formula/liblcf.rb#L42-L43) not suffer from the same problem?

If they do, the regex as it is won’t catch them. If they don’t, then modifying `kafka` to conform to their pattern seems like a better solution, as it seems to be the outlier.

Reply:
Thanks for your comments guys. I did think about the RuboCop option and after your pointers I reckon that **would** be the better option
* remove the check from the FormulaAuditor#line_problems method
* use the AST parsed methods in RuboCop instead.

It's a good chance to learn a bit more RuboCop :-) Leave it with me.

Reply:
Sounds good, thanks!